File: dinner_planner/lib/models/pantry_item.dart
class PantryItem {
  final int? id;
  final String name;
  double quantity;
  final String unit;
  final String? barcode;
  final double? calories;
  final double? protein;
  final double? carbs;
  final double? fat;
  final DateTime? expirationDate;
  final DateTime createdAt;

  PantryItem({
    this.id,
    required this.name,
    required this.quantity,
    required this.unit,
    this.barcode,
    this.calories,
    this.protein,
    this.carbs,
    this.fat,
    this.expirationDate,
    DateTime? createdAt,
  }) : createdAt = createdAt ?? DateTime.now();

  bool get isExpired =>
      expirationDate != null &&
      expirationDate!.isBefore(DateTime.now());

  bool get isExpiringSoon {
    if (expirationDate == null) return false;
    final daysLeft =
        expirationDate!.difference(DateTime.now()).inDays;
    return daysLeft <= 2 && !isExpired;
  }

  factory PantryItem.fromJson(Map<String, dynamic> json) {
    return PantryItem(
      id: json['id'] as int?,
      name: json['name'] as String,
      quantity: (json['quantity'] as num).toDouble(),
      unit: json['unit'] as String,
      barcode: json['barcode'] as String?,
      calories: (json['calories'] as num?)?.toDouble(),
      protein: (json['protein'] as num?)?.toDouble(),
      carbs: (json['carbs'] as num?)?.toDouble(),
      fat: (json['fat'] as num?)?.toDouble(),
      expirationDate: json['expiration_date'] != null
          ? DateTime.tryParse(json['expiration_date'] as String)
          : null,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() => {
        'name': name,
        'quantity': quantity,
        'unit': unit,
        if (barcode != null) 'barcode': barcode,
        if (calories != null) 'calories': calories,
        if (protein != null) 'protein': protein,
        if (carbs != null) 'carbs': carbs,
        if (fat != null) 'fat': fat,
        if (expirationDate != null)
          'expiration_date':
              '${expirationDate!.year.toString().padLeft(4, '0')}-'
              '${expirationDate!.month.toString().padLeft(2, '0')}-'
              '${expirationDate!.day.toString().padLeft(2, '0')}',
      };
}

File: web/src/types/pantryItem.ts
export interface PantryItem {
  id?: number
  name: string
  quantity: number
  unit: string
  barcode?: string | null
  calories?: number | null
  protein?: number | null
  carbs?: number | null
  fat?: number | null
  expirationDate?: string | null // 'YYYY-MM-DD'
  createdAt: string
}

export function isExpired(item: PantryItem): boolean {
  if (!item.expirationDate) return false
  return new Date(item.expirationDate) < new Date(new Date().toDateString())
}

export function isExpiringSoon(item: PantryItem): boolean {
  if (!item.expirationDate) return false
  const today = new Date(new Date().toDateString())
  const exp = new Date(item.expirationDate)
  const daysLeft = Math.round((exp.getTime() - today.getTime()) / (1000 * 60 * 60 * 24))
  return daysLeft <= 2 && !isExpired(item)
}

export function pantryItemFromJson(json: any): PantryItem {
  return {
    id: json.id,
    name: json.name,
    quantity: Number(json.quantity ?? 0),
    unit: json.unit,
    barcode: json.barcode ?? null,
    calories: json.calories != null ? Number(json.calories) : null,
    protein: json.protein != null ? Number(json.protein) : null,
    carbs: json.carbs != null ? Number(json.carbs) : null,
    fat: json.fat != null ? Number(json.fat) : null,
    expirationDate: json.expiration_date ?? null,
    createdAt: json.created_at ?? new Date().toISOString(),
  }
}

export function pantryItemToJson(item: PantryItem): Record<string, unknown> {
  return {
    name: item.name,
    quantity: item.quantity,
    unit: item.unit,
    ...(item.barcode != null ? { barcode: item.barcode } : {}),
    ...(item.calories != null ? { calories: item.calories } : {}),
    ...(item.protein != null ? { protein: item.protein } : {}),
    ...(item.carbs != null ? { carbs: item.carbs } : {}),
    ...(item.fat != null ? { fat: item.fat } : {}),
    ...(item.expirationDate ? { expiration_date: item.expirationDate } : {}),
  }
}
